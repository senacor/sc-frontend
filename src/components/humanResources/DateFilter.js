import React, { Component } from 'react';
import * as actions from '../../actions';
import { getSubFilter } from '../../reducers/selector';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ListItem from '@material-ui/core/ListItem/ListItem';
import List from '@material-ui/core/List/List';
import TextField from '@material-ui/core/TextField/TextField';
import IconButton from '@material-ui/core/IconButton/IconButton';
import Icon from '@material-ui/core/Icon/Icon';
import moment from 'moment';
import { withStyles } from '@material-ui/core';

let styles = {
  error: {
    font: 'small',
    color: '#ff0000'
  }
};

export class DateFilter extends Component {
  constructor(props) {
    super(props);
    const filter = Object.assign(
      {},
      { searchString: '', values: '' },
      this.props.filter
    );

    this.state = {
      filter: filter,
      values: filter.values,
      error: false
    };
  }

  showError = () => {
    this.setState({ error: true });
  };

  isValidFilter = values => {
    return moment(values.From).isBefore(moment(values.To));
  };

  setFilter = values => {
    let searchString = Object.keys(values)
      .map(function(key) {
        return `${this.props.filterBy}${key}=${values[key]}`;
      }, this)
      .join('&');

    let filter = {
      searchString,
      values: values
    };
    let payload = {
      filterGroup: this.props.filterGroup,
      filterBy: this.props.filterBy,

      filter: filter
    };

    if (searchString === '') {
      this.props.deleteFilter(payload);
    } else if (!this.isValidFilter(values)) {
      return false;
    } else {
      this.props.addFilter(payload);
    }
    return true;
  };

  onChange = key => event => {
    let newValues = Object.assign({}, this.state.values, {
      [key]: event.target.value
    });
    if (newValues[key] === '') {
      delete newValues[key];
    }

    this.setState({ values: newValues, error: false });
  };

  execute = () => {
    if (this.setFilter(this.state.values)) {
      this.props.closeFilter();
    } else {
      this.showError();
    }
  };

  render() {
    const { classes } = this.props;
    let { values } = this.state;
    return (
      <List>
        <ListItem>
          <TextField
            id="startDate"
            label="von"
            type="date"
            defaultValue={values.From}
            InputLabelProps={{
              shrink: true
            }}
            onChange={this.onChange('From')}
          />
          <TextField
            id="endDate"
            label="bis"
            type="date"
            defaultValue={values.To}
            InputLabelProps={{
              shrink: true
            }}
            onChange={this.onChange('To')}
          />
          <IconButton
            id="forwardButton"
            aria-label="forward"
            onClick={this.execute}
          >
            <Icon>forward</Icon>
          </IconButton>
        </ListItem>
        <ListItem className={classes.error}>
          {this.state.error === true ? (
            <div>Falsche Eingabe: von nach bis</div>
          ) : null}
        </ListItem>
      </List>
    );
  }
}

DateFilter.propTypes = {
  filterGroup: PropTypes.string.isRequired,
  filterBy: PropTypes.string.isRequired
};

export const StyledComponent = withStyles(styles)(DateFilter);
export default connect(
  (state, props) => ({
    filter: getSubFilter(props.filterGroup, props.filterBy)(state)
  }),
  {
    addFilter: actions.addFilter,
    deleteFilter: actions.deleteFilter
  }
)(StyledComponent);
