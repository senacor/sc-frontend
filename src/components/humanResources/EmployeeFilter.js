import React, { Component } from 'react';
import * as actions from '../../actions';
import EmployeeSearch from '../employeeSearch/EmployeeSearch';
import { getFilter } from '../../reducers/selector';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

class EmployeeFilter extends Component {
  constructor(props) {
    super(props);

    let filter;
    let emptyFilter = {
      [this.props.filterBy]: { searchString: '', values: '' }
    };
    if (!this.props.filter || !this.props.filter[this.props.filterGroup]) {
      filter = emptyFilter;
    } else {
      filter = Object.assign(
        {},
        emptyFilter,
        this.props.filter[this.props.filterGroup]
      );
    }

    this.state = {
      employeeName: filter[this.props.filterBy].values,
      filter: filter
    };
  }

  selectEmployee = employee => {
    if (employee) {
      let filter = this.state.filter;
      filter[this.props.filterBy] = {
        searchString: `${this.props.filterBy}=${employee.id}`,
        values: `${employee.firstName} ${employee.lastName}`
      };
      let payload = {
        filterGroup: this.props.filterGroup,
        filter: filter
      };
      this.props.fetchFilteredPrsForHumanResource(payload);
    }
  };

  render() {
    return (
      <EmployeeSearch
        employeeSearchValue={this.state.employeeName}
        selectEmployee={this.selectEmployee}
      />
    );
  }
}

EmployeeFilter.propTypes = {
  filterGroup: PropTypes.string.isRequired,
  filterBy: PropTypes.string.isRequired
};

export default connect(
  state => ({
    filter: getFilter(state)
  }),
  {
    fetchFilteredPrsForHumanResource: actions.fetchFilteredPrsForHumanResource
  }
)(EmployeeFilter);
