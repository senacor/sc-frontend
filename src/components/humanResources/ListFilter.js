import React, { Component } from 'react';
import * as actions from '../../actions';
import { getSubFilter } from '../../reducers/selector';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Checkbox from '@material-ui/core/Checkbox/Checkbox';
import ListItemText from '@material-ui/core/ListItemText/ListItemText';
import ListItem from '@material-ui/core/ListItem/ListItem';
import List from '@material-ui/core/List/List';
import Divider from '@material-ui/core/Divider/Divider';
import Icon from '@material-ui/core/Icon/Icon';
import {Button} from "@material-ui/core";

export class ListFilter extends Component {
  constructor(props) {
    super(props);

    const filter = Object.assign(
      {},
      { searchString: '', values: '' },
      this.props.filter
    );
    const checked =
      filter.values === '' ? Object.keys(props.content) : filter.values;
    this.state = {
      filter: filter,
      checked: checked,
      isAllSelected: checked.length === Object.keys(props.content).length
    };
  }

  isFilterSet = checked => {
    return !(checked.length === Object.keys(this.props.content).length ||
              checked.length === 0);
  };

  isFilterEmpty = checked => {
    return checked.length === 0;
  }

  handleToggleSelectAll = () => {
    const newAllSelect = !this.state.isAllSelected;
    let newChecked;

    if (newAllSelect === true) {
      newChecked = Object.keys(this.props.content);
    } else {
      newChecked = [];
    }
    this.setState({
      isAllSelected: newAllSelect,
      checked: newChecked
    });
  };

  handleToggle = value => () => {
    const { checked } = this.state;
    const currentIndex = checked.indexOf(value);
    const newChecked = [...checked];

    if (currentIndex === -1) {
      newChecked.push(value);
    } else {
      newChecked.splice(currentIndex, 1);
    }

    if (this.isFilterSet(newChecked)) {
      // something is set in filter, but not everything
      this.setState({
        checked: newChecked,
        isAllSelected: false
      });
    } else if (this.isFilterEmpty(newChecked)) {
      // nothing is set in filter
      console.log('new', newChecked)
      this.setState({
        checked: newChecked,
        isAllSelected: false
      });
    } else {
      // everything is set in filter
      this.setState({
        checked: Object.keys(this.props.content),
        isAllSelected: true
      });
    }
  };

  setFilter = () => {
    let searchString = `${this.props.filterBy}=`;
    searchString += this.state.checked
      .map(value => {
        return this.props.content[value];
      })
      .join(`&${this.props.filterBy}=`);

    let filter = {
      searchString,
      values: this.state.checked
    };
    let payload = {
      filterGroup: this.props.filterGroup,
      filterBy: this.props.filterBy,

      filter: filter
    };
    this.props.addFilter(payload);
  };

  showSelectAll = () => {
    return (
      <ListItem
        key={'selectAll'}
        dense
        button
        onClick={this.handleToggleSelectAll}
      >
        <Checkbox
          id={'selectAll'}
          checked={this.state.isAllSelected}
          color={'primary'}
          checkedIcon={<Icon>check</Icon>}
          icon={<Icon />}
        />
        <ListItemText primary={'selectAll'} />
      </ListItem>
    );
  };
  showContent = value => {
    return (
      <ListItem key={value} dense button onClick={this.handleToggle(value)}>
        <Checkbox
          id={value}
          checked={this.state.checked.indexOf(value) !== -1}
          color={'primary'}
          checkedIcon={<Icon>check</Icon>}
          icon={<Icon />}
        />
        <ListItemText primary={value} />
      </ListItem>
    );
  };

  render() {
    return (
      <List>
        {this.showSelectAll()}
        <Divider />
        {Object.keys(this.props.content).map(this.showContent, this)}
        <Divider />
        <Button onClick={this.setFilter}
                style={{backgroundColor: '#06A781', float: 'right'}}
        >
          OK
        </Button>
      </List>

    );
  }
}

ListFilter.propTypes = {
  filterGroup: PropTypes.string.isRequired,
  filterBy: PropTypes.string.isRequired,
  content: PropTypes.object.isRequired
};

export default connect(
  (state, props) => ({
    filter: getSubFilter(props.filterGroup, props.filterBy)(state)
  }),
  {
    addFilter: actions.addFilter,
    deleteFilter: actions.deleteFilter
  }
)(ListFilter);
