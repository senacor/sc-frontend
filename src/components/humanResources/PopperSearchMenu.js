import React, { useState } from 'react';
import Popover from '@material-ui/core/Popover';
import Icon from '@material-ui/core/Icon/Icon';
import IconButton from '@material-ui/core/IconButton/IconButton';
import { getSubFilter } from '../../reducers/selector';
import { connect } from 'react-redux';
import { withStyles } from '@material-ui/core';

const styles = theme => ({
  iconFilterSet: {
    color: theme.palette.primary['500'],
    margin: '0px'
  },
  iconFilterUnset: {
    color: '#dddddd',
    margin: '0px'
  }
});

export const PopperSearchMenu = ({
  classes,
  subfilter,
  children,
  filterBy,
  filterGroup
}) => {
  const [anchorEl, setAnchorEl] = useState(null);

  const handleClick = event => {
    setAnchorEl(event.currentTarget);
    event.stopPropagation();
  };

  const handleClose = event => {
    setAnchorEl(null);
  };

  const getIcon = subfilter => {
    return subfilter === '' ? (
      <Icon className={classes.iconFilterUnset}>filter_list</Icon>
    ) : (
      <Icon className={classes.iconFilterSet}>filter_list</Icon>
    );
  };

  const open = Boolean(anchorEl);
  return (
    <div>
      <IconButton onClick={handleClick}>{getIcon(subfilter)}</IconButton>
      <Popover
        id="simple-popper"
        open={open}
        anchorEl={anchorEl}
        onClose={handleClose}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center'
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'center'
        }}
        onClick={event => {
          event.stopPropagation();
        }}
      >
        <div>
          {React.cloneElement(children, {
            closeFilter: handleClose,
            filterBy: filterBy,
            filterGroup: filterGroup
          })}
        </div>
      </Popover>
    </div>
  );
};

export const StyledComponent = withStyles(styles)(PopperSearchMenu);
export default connect(
  (state, props) => ({
    subfilter: getSubFilter(props.filterGroup, props.filterBy)(state)
  }),
  {}
)(StyledComponent);
