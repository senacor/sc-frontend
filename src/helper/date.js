import moment from 'moment';
import 'moment-timezone';

export const FRONTEND_DATE_FORMAT = 'DD.MM.YYYY';
export const FRONTEND_LOCALE_DATE_TIME_FORMAT = 'YYYY-MM-DD';

export const formatDateForFrontend = input => {
  if (input) {
    let date = moment.isMoment(input) ? input : moment(input);
    if (date.isValid()) {
      return date.format(FRONTEND_DATE_FORMAT);
    }
  }
  return null;
};

// Converting LocaleDateTime for frontend
export const formatLocaleDateTime = (input, format) => {
  if (input) {
    let modifiedInput = [...input];
    modifiedInput[1] = modifiedInput[1] - 1;
    let date = moment.isMoment(modifiedInput)
      ? modifiedInput
      : moment.utc(modifiedInput);
    if (date.isValid()) {
      return date.tz('Europe/Berlin').format(format);
    }
  }
  return null;
};

// Converting date to LocaleDateTime for backend
export const convertLocalDateTime = date => {
  const convertedDate = moment
    .tz(`${date.target.value}`, 'Europe/Berlin')
    .utc()
    .format('YYYY-MM-DDTHH:mm');
  if (convertedDate === 'Invalid date') {
    return null;
  } else return convertedDate;
};

export const isValidDate = date => {
  const isValidDate = moment(
    date,
    FRONTEND_LOCALE_DATE_TIME_FORMAT,
    true
  ).isValid();
  return isValidDate;
};

export const getReadableDate = date => {
  const minutes = date[4] > 9 ? date[4].toString() : `0${date[4]}`;
  return `${date[2]}.${date[1]}.${date[0]}, ${date[3]}:${minutes}`;
};
