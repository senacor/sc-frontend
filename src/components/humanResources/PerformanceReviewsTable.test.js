import React from 'react';
import PerformanceReviewsTable from './PerformanceReviewsTable';
import { descInteger, descString, stableSort } from './PerformanceReviewsTable';
import { createShallow } from '@material-ui/core/test-utils';
import HR_ELEMENTS from './hrElements';
import { Link } from 'react-router-dom';
import PopperSearchMenu from './PopperSearchMenu';
import EmployeeFilter from './EmployeeFilter';
import { translateContent } from '../translate/Translate';
import FILTER_GROUPS from './filterGroups';
import getDisplayName from '../../helper/getDisplayName';
import { formatDateForFrontend } from '../../helper/date';
import DateFilter from './DateFilter';
import ListFilter from './ListFilter';

const rows = [
  {
    numeric: false,
    disablePadding: false,
    label: 'Mitarbeiter',
    sortValue: entry => getDisplayName(entry[HR_ELEMENTS.EMPLOYEE]),
    render: entry => {
      return (
        <Link to={`/prDetail/${entry.prId}`}>
          {getDisplayName(entry[HR_ELEMENTS.EMPLOYEE])}
        </Link>
      );
    },
    filter: (
      <PopperSearchMenu
        filterGroup={FILTER_GROUPS.HR}
        filterBy={HR_ELEMENTS.EMPLOYEE}
      >
        <EmployeeFilter />
      </PopperSearchMenu>
    )
  },
  {
    numeric: true,
    disablePadding: true,
    label: 'Fälligkeit',
    sortValue: entry => entry[HR_ELEMENTS.DEADLINE],
    render: entry => formatDateForFrontend(entry[HR_ELEMENTS.DEADLINE]),
    filter: (
      <PopperSearchMenu
        filterGroup={FILTER_GROUPS.HR}
        filterBy={HR_ELEMENTS.DEADLINE}
      >
        <DateFilter />
      </PopperSearchMenu>
    )
  },
  {
    numeric: false,
    disablePadding: false,
    label: 'Grund',
    sortValue: entry => translateContent(entry[HR_ELEMENTS.PR_OCCASION]),
    render: entry => translateContent(entry[HR_ELEMENTS.PR_OCCASION]),
    filter: (
      <PopperSearchMenu filterGroup={FILTER_GROUPS.HR} filterBy={'occasion'}>
        <ListFilter
          content={{
            [translateContent('ON_DEMAND')]: 'ON_DEMAND',
            [translateContent('END_PROBATION')]: 'END_PROBATION',
            [translateContent('YEARLY')]: 'YEARLY'
          }}
        />
      </PopperSearchMenu>
    )
  },
  {
    numeric: false,
    disablePadding: false,
    label: 'Projektkst',
    sortValue: entry => entry[HR_ELEMENTS.CST],
    render: entry => entry[HR_ELEMENTS.CST]
  },
  {
    numeric: false,
    disablePadding: true,
    label: 'Dev/Con',
    sortValue: entry =>
      translateContent(`COMPETENCE_${entry[HR_ELEMENTS.COMPETENCE]}`),
    render: entry =>
      translateContent(`COMPETENCE_${entry[HR_ELEMENTS.COMPETENCE]}`),
    filter: (
      <PopperSearchMenu
        filterGroup={FILTER_GROUPS.HR}
        filterBy={HR_ELEMENTS.COMPETENCE}
      >
        <ListFilter
          content={{
            [translateContent('COMPETENCE_DEVELOPMENT')]: 'DEVELOPMENT',
            [translateContent('COMPETENCE_CONSULTING')]: 'CONSULTING'
          }}
        />
      </PopperSearchMenu>
    )
  },
  {
    numeric: true,
    disablePadding: true,
    label: 'level',
    sortValue: entry => entry[HR_ELEMENTS.LEVEL],
    render: entry => entry[HR_ELEMENTS.LEVEL],
    filter: (
      <PopperSearchMenu
        filterGroup={FILTER_GROUPS.HR}
        filterBy={HR_ELEMENTS.LEVEL}
      >
        <ListFilter
          content={{
            1: 1,
            2: 2
          }}
        />
      </PopperSearchMenu>
    )
  },
  {
    numeric: false,
    disablePadding: true,
    label: 'Vorgesetzte/r',
    sortValue: entry => getDisplayName(entry[HR_ELEMENTS.SUPERVISOR]),
    render: entry => getDisplayName(entry[HR_ELEMENTS.SUPERVISOR]),
    filter: (
      <PopperSearchMenu
        filterGroup={FILTER_GROUPS.HR}
        filterBy={HR_ELEMENTS.SUPERVISOR}
      >
        <EmployeeFilter />
      </PopperSearchMenu>
    )
  },
  {
    numeric: false,
    disablePadding: true,
    label: 'Bewerter',
    sortValue: entry => getDisplayName(entry[HR_ELEMENTS.REVIEWER]),
    render: entry => getDisplayName(entry[HR_ELEMENTS.REVIEWER]),
    filter: (
      <PopperSearchMenu
        filterGroup={FILTER_GROUPS.HR}
        filterBy={HR_ELEMENTS.REVIEWER}
      >
        <EmployeeFilter />
      </PopperSearchMenu>
    )
  },
  {
    numeric: false,
    disablePadding: true,
    label: 'Bewertung',
    sortValue: entry => entry[HR_ELEMENTS.RESULT],
    render: entry => entry[HR_ELEMENTS.RESULT]
  },
  {
    numeric: false,
    disablePadding: true,
    label: 'MA ausgefüllt',
    sortValue: entry =>
      entry[HR_ELEMENTS.EMPLOYEE_PREPARATION_DONE] ? 'ja' : 'nein',
    render: entry =>
      entry[HR_ELEMENTS.EMPLOYEE_PREPARATION_DONE] ? 'ja' : 'nein',
    filter: (
      <PopperSearchMenu
        filterGroup={FILTER_GROUPS.HR}
        filterBy={'isEmployeePreparationDone'}
      >
        <ListFilter content={{ ja: true, nein: false }} />
      </PopperSearchMenu>
    )
  },
  {
    numeric: false,
    disablePadding: false,
    label: 'Beurteiler ausgefüllt',
    sortValue: entry =>
      entry[HR_ELEMENTS.REVIEWER_PREPARATION_DONE] ? 'ja' : 'nein',
    render: entry =>
      entry[HR_ELEMENTS.REVIEWER_PREPARATION_DONE] ? 'ja' : 'nein',
    filter: (
      <PopperSearchMenu
        filterGroup={FILTER_GROUPS.HR}
        filterBy={'isReviewerPreparationDone'}
      >
        <ListFilter content={{ ja: true, nein: false }} />
      </PopperSearchMenu>
    )
  },
  {
    key: HR_ELEMENTS.APPOINTMENT,
    numeric: false,
    disablePadding: true,
    label: 'Termin',
    sortValue: entry => entry[HR_ELEMENTS.APPOINTMENT],
    render: entry => entry[HR_ELEMENTS.APPOINTMENT]
  },
  {
    numeric: false,
    disablePadding: true,
    label: 'Finaler Status',
    sortValue: entry =>
      entry[HR_ELEMENTS.IN_PROGRESS] ? 'laufend' : 'abgeschlossen',
    render: entry =>
      entry[HR_ELEMENTS.IN_PROGRESS] ? 'laufend' : 'abgeschlossen',
    filter: (
      <PopperSearchMenu
        filterGroup={FILTER_GROUPS.HR}
        filterBy={'isInProgress'}
      >
        <ListFilter content={{ laufend: true, abgeschlossen: false }} />
      </PopperSearchMenu>
    )
  },
  {
    numeric: false,
    disablePadding: true,
    label: 'HR verarbeitet',
    sortValue: entry => (entry[HR_ELEMENTS.HR_PROCESSING_DONE] ? 'ja' : 'nein'),
    render: entry => (entry[HR_ELEMENTS.HR_PROCESSING_DONE] ? 'ja' : 'nein'),
    filter: (
      <PopperSearchMenu
        filterGroup={FILTER_GROUPS.HR}
        filterBy={'isHumanResourceProcessingDone'}
      >
        <ListFilter content={{ ja: true, nein: false }} />
      </PopperSearchMenu>
    )
  }
];

describe('PerformanceReviewsTable component', () => {
  let shallow = createShallow({ dive: true });

  it('should match snapshot', () => {
    let prs = [
      getTestdataEntry(1, '2018-01-01', 'DEVELOPMENT', 'ON_DEMAND'),
      getTestdataEntry(2, '2018-02-01', 'CONSULTING', 'ON_DEMAND')
    ];

    let component = shallow(
      <PerformanceReviewsTable data={prs} columnDefinition={rows} orderBy={1} />
    );

    expect(component).toMatchSnapshot();
  });
});

describe('descInteger', () => {
  let prs = [
    getTestdataEntry(1, '2018-01-01', 'DEVELOPMENT', 'ON_DEMAND'),
    getTestdataEntry(2, '2018-01-01', 'CONSULTING', 'ON_DEMAND'),
    getTestdataEntry(3, '2018-02-10', 'CONSULTING', 'ON_DEMAND'),
    getTestdataEntry(4, '2018-01-11', 'CONSULTING', 'ON_DEMAND')
  ];

  it('should return 0: equal', () => {
    const result = descInteger(prs[0], prs[1], entity => entity['deadline']);
    expect(result).toEqual(0);
  });
  it('should return 1: asc', () => {
    const result = descInteger(prs[1], prs[2], entity => entity['deadline']);
    expect(result).toEqual(1);
  });
  it('should return -1: desc', () => {
    const result = descInteger(prs[2], prs[3], entity => entity['deadline']);
    expect(result).toEqual(-1);
  });
});

describe('descString', () => {
  let prs = [
    getTestdataEntry(1, '2018-01-01', 'CONSULTING', 'ON_DEMAND'),
    getTestdataEntry(2, '2018-01-01', 'CONSULTING', 'ON_DEMAND'),
    getTestdataEntry(3, '2018-02-10', 'DEVELOPMENT', 'ON_DEMAND'),
    getTestdataEntry(4, '2018-01-11', 'CONSULTING', 'ON_DEMAND')
  ];

  it('should return 0: equal', () => {
    const result = descString(prs[0], prs[1], entity => entity['competence']);
    expect(result).toBe(0);
  });
  it('should return 1: asc', () => {
    const result = descString(prs[1], prs[2], entity => entity['competence']);
    expect(result).toEqual(1);
  });
  it('should return -1: desc', () => {
    const result = descString(prs[2], prs[3], entity => entity['competence']);
    expect(result).toEqual(-1);
  });
});

describe('stableSort', () => {
  let prs = [
    getTestdataEntry(1, '2018-02-01', 'DEVELOPMENT', 'ON_DEMAND'),
    getTestdataEntry(2, '2018-01-11', 'CONSULTING', 'ON_DEMAND'),
    getTestdataEntry(3, '2018-02-10', 'CONSULTING', 'ON_DEMAND'),
    getTestdataEntry(4, '2019-01-01', 'CONSULTING', 'ON_DEMAND')
  ];

  it('should return a asc array', () => {
    const result = stableSort(prs, (a, b) =>
      descInteger(a, b, entry => entry['deadline'])
    );
    expect(result[0].prId).toEqual(4);
    expect(result[1].prId).toEqual(3);
    expect(result[2].prId).toEqual(1);
    expect(result[3].prId).toEqual(2);
  });

  it('should return a desc array', () => {
    const result = stableSort(
      prs,
      (a, b) => -descInteger(a, b, entry => entry['deadline'])
    );
    expect(result[0].prId).toEqual(2);
    expect(result[1].prId).toEqual(1);
    expect(result[2].prId).toEqual(3);
    expect(result[3].prId).toEqual(4);
  });
});

function getTestdataEntry(id, deadline, competence, occasion) {
  return {
    prId: id,
    employee: {
      id: 502,
      firstName: 'Michaela',
      lastName: 'Mitarbeiterin',
      login: 'test.pr.mitarbeiter1',
      dateOfLastPr: null,
      correspondingEventResponses: null
    },
    deadline: deadline,
    prOccasion: occasion,
    projectCst: null,
    competence: competence,
    level: 1,
    supervisor: {
      id: 501,
      firstName: 'Volker',
      lastName: 'Vorgesetzter',
      login: 'test.pr.vorgesetzter',
      dateOfLastPr: null,
      correspondingEventResponses: null
    },
    reviewer: {
      id: 501,
      firstName: 'Volker',
      lastName: 'Vorgesetzter',
      login: 'test.pr.vorgesetzter',
      dateOfLastPr: null,
      correspondingEventResponses: null
    },
    appointment: null,
    employeePreparationDone: true,
    reviewerPreparationDone: false,
    humanResourceProcessingDone: false,
    inProgress: false,
    _links: {
      self: {
        href: 'http://localhost:8010/api/v1/prs/4'
      }
    }
  };
}
