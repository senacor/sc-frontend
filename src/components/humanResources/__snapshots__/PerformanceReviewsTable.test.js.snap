// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PerformanceReviewsTable component should match snapshot 1`] = `
<WithStyles(Paper)
  className="PerformanceReviewsTable-root-1"
>
  <div
    className="PerformanceReviewsTable-tableWrapper-3"
  >
    <WithStyles(Table)
      aria-labelledby="tableTitle"
      className="PerformanceReviewsTable-table-2"
    >
      <EnhancedTableHead
        columnDefinition={
          Array [
            Object {
              "disablePadding": false,
              "filter": <PopperSearchMenu>
                <Connect(EmployeeFilter)
                      filterBy="employee"
                      filterGroup="hr"
                />
          </PopperSearchMenu>,
              "key": "employee",
              "label": "Mitarbeiter",
              "mapper": [Function],
              "numeric": false,
              "show": [Function],
            },
            Object {
              "disablePadding": true,
              "filter": <PopperSearchMenu
                content={
                      <WithStyles(List)>
                            <WithStyles(ListItem)>
                                  <TextField
                                        InputLabelProps={
                                              Object {
                                                    "shrink": true,
                                                  }
                                        }
                                        id="dateBegin"
                                        label="Fälligkeit von"
                                        required={false}
                                        select={false}
                                        type="date"
                                  />
                                  <TextField
                                        InputLabelProps={
                                              Object {
                                                    "shrink": true,
                                                  }
                                        }
                                        defaultValue="2017-05-24"
                                        id="dateEnd"
                                        label="Fälligkeit bis"
                                        required={false}
                                        select={false}
                                        type="date"
                                  />
                            </WithStyles(ListItem)>
                      </WithStyles(List)>
                }
          />,
              "key": "deadline",
              "label": "Fälligkeit",
              "mapper": [Function],
              "numeric": true,
              "show": [Function],
            },
            Object {
              "disablePadding": false,
              "key": "prOccasion",
              "label": "Grund",
              "mapper": [Function],
              "numeric": false,
              "show": [Function],
            },
            Object {
              "disablePadding": false,
              "key": "projectCst",
              "label": "Projektkst",
              "mapper": [Function],
              "numeric": false,
              "show": [Function],
            },
            Object {
              "disablePadding": true,
              "key": "competence",
              "label": "Dev/Con",
              "mapper": [Function],
              "numeric": false,
              "show": [Function],
            },
            Object {
              "disablePadding": true,
              "key": "level",
              "label": "level",
              "mapper": [Function],
              "numeric": false,
              "show": [Function],
            },
            Object {
              "disablePadding": true,
              "filter": <PopperSearchMenu>
                <Connect(EmployeeFilter)
                      filterBy="supervisor"
                      filterGroup="hr"
                />
          </PopperSearchMenu>,
              "key": "supervisor",
              "label": "Vorgesetzte/r",
              "mapper": [Function],
              "numeric": false,
              "show": [Function],
            },
            Object {
              "disablePadding": true,
              "filter": <PopperSearchMenu>
                <Connect(EmployeeFilter)
                      filterBy="reviewer"
                      filterGroup="hr"
                />
          </PopperSearchMenu>,
              "key": "reviewer",
              "label": "Bewerter",
              "mapper": [Function],
              "numeric": false,
              "show": [Function],
            },
            Object {
              "disablePadding": true,
              "key": "result",
              "label": "Bewertung",
              "mapper": [Function],
              "numeric": false,
              "show": [Function],
            },
            Object {
              "disablePadding": true,
              "key": "employeePreparationDone",
              "label": "MA ausgefüllt",
              "mapper": [Function],
              "numeric": false,
              "show": [Function],
            },
            Object {
              "disablePadding": false,
              "key": "reviewerPreparationDone",
              "label": "Beurteiler ausgefüllt",
              "mapper": [Function],
              "numeric": false,
              "show": [Function],
            },
            Object {
              "disablePadding": true,
              "key": "appointment",
              "label": "Termin",
              "mapper": [Function],
              "numeric": false,
              "show": [Function],
            },
            Object {
              "disablePadding": true,
              "key": "inProgress",
              "label": "Finaler Status",
              "mapper": [Function],
              "numeric": false,
              "show": [Function],
            },
            Object {
              "disablePadding": true,
              "key": "humanResourceProcessingDone",
              "label": "HR verarbeitet",
              "mapper": [Function],
              "numeric": false,
              "show": [Function],
            },
          ]
        }
        onRequestSort={[Function]}
        order="desc"
        orderBy="deadline"
      />
      <WithStyles(TableBody)>
        <WithStyles(TableRow)
          hover={true}
          key="0"
          tabIndex={-1}
        >
          <WithStyles(TableCell)
            key="employee"
            padding="default"
          >
            <Link
              replace={false}
              to="/prDetail/2"
            >
              Michaela Mitarbeiterin
            </Link>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="deadline"
            padding="default"
          >
            2018-02-01
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="prOccasion"
            padding="default"
          >
            <Translate
              content="ON_DEMAND"
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="projectCst"
            padding="default"
          />
          <WithStyles(TableCell)
            key="competence"
            padding="default"
          >
            <Translate
              content="COMPETENCE_CONSULTANT"
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="level"
            padding="default"
          >
            1
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="supervisor"
            padding="default"
          >
            Volker Vorgesetzter
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="reviewer"
            padding="default"
          >
            Volker Vorgesetzter
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="result"
            padding="default"
          />
          <WithStyles(TableCell)
            key="employeePreparationDone"
            padding="default"
          >
            ja
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="reviewerPreparationDone"
            padding="default"
          >
            nein
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="appointment"
            padding="default"
          />
          <WithStyles(TableCell)
            key="inProgress"
            padding="default"
          >
            abgeschlossen
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="humanResourceProcessingDone"
            padding="default"
          >
            nein
          </WithStyles(TableCell)>
        </WithStyles(TableRow)>
        <WithStyles(TableRow)
          hover={true}
          key="1"
          tabIndex={-1}
        >
          <WithStyles(TableCell)
            key="employee"
            padding="default"
          >
            <Link
              replace={false}
              to="/prDetail/1"
            >
              Michaela Mitarbeiterin
            </Link>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="deadline"
            padding="default"
          >
            2018-01-01
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="prOccasion"
            padding="default"
          >
            <Translate
              content="ON_DEMAND"
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="projectCst"
            padding="default"
          />
          <WithStyles(TableCell)
            key="competence"
            padding="default"
          >
            <Translate
              content="COMPETENCE_DEVELOPMENT"
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="level"
            padding="default"
          >
            1
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="supervisor"
            padding="default"
          >
            Volker Vorgesetzter
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="reviewer"
            padding="default"
          >
            Volker Vorgesetzter
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="result"
            padding="default"
          />
          <WithStyles(TableCell)
            key="employeePreparationDone"
            padding="default"
          >
            ja
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="reviewerPreparationDone"
            padding="default"
          >
            nein
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="appointment"
            padding="default"
          />
          <WithStyles(TableCell)
            key="inProgress"
            padding="default"
          >
            abgeschlossen
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="humanResourceProcessingDone"
            padding="default"
          >
            nein
          </WithStyles(TableCell)>
        </WithStyles(TableRow)>
      </WithStyles(TableBody)>
    </WithStyles(Table)>
  </div>
  <WithStyles(TablePagination)
    backIconButtonProps={
      Object {
        "aria-label": "Previous Page",
      }
    }
    component="div"
    count={2}
    nextIconButtonProps={
      Object {
        "aria-label": "Next Page",
      }
    }
    onChangePage={[Function]}
    onChangeRowsPerPage={[Function]}
    page={0}
    rowsPerPage={25}
  />
</WithStyles(Paper)>
`;
